spec/unit/util_spec.rb

```diff
@@ -871,13 +871,13 @@ def withenv_utf8(&block)
       end
     end
 
-    it "should copy the permissions of the source file before yielding on Unix", :if => !Puppet.features.microsoft_windows? do
+    it "should copy the permissions of the source file after yielding on Unix", :if => !Puppet.features.microsoft_windows? do
       set_mode(0555, target.path)
       inode = Puppet::FileSystem.stat(target.path).ino
 
       yielded = false
-      subject.replace_file(target.path, 0600) do |fh|
-        expect(get_mode(fh.path)).to eq(0555)
+      subject.replace_file(target.path, 0660) do |fh|
+        expect(get_mode(fh.path)).to eq(0600)
```

Comment:
Hmm... shouldn't the assertion here be that the mode is `0660` as requested?

```ruby
expect(get_mode(fh.path)).to eq(0660)
```

Do we know why the mode isn't set the same as requested?

Reply:
`0660` is the requested final permissions if the file doesn't already exist. Inside the block, the permissions should be `0600` as that's the "secure" permissions used for creating temp files (which also still allow us to write to the file). The permissions are changed to the final permissions after the block is executed as the requested permissions might be read-only.

Reply:
Ah yes, this is *in the block*... I understand now.  Sorry I misinterpreted that bit.

And the behavior when the file exists is to completely ignore the `default_mode` passed as `0660`, instead resetting to the perms on the original file (in this case `0555` setup earlier)?

I think Moses and I had originally discussed that that part is also a bit surprising / weird, but given that's what we did in the past, that's OK. At least this PR fixes the write-ability in the block problem!

Reply:
Your description of `default_mode` is correct. It's only used if we don't have a pre-existing file to copy the permissions from.